sympy/combinatorics/fp_groups.py

```diff
@@ -259,6 +286,61 @@ def scan(self, alpha, word):
             self.table[b][A_dict_inv[word[i]]] = f
         # otherwise scan is incomplete and yields no information
 
+    def scan_check(self, alpha, word):
+        """
+        Another version of "scan" routine, it checks whether α scans correctly
+        under w, it is a straightforward modification of "scan". "scan_check"
+        return false (rather than calling "coincidence") if the scan completes
+        incorrectly; otherwise it returns true.
+
+        Examples
+        ========
+
+        >>> from sympy.combinatorics.fp_groups import FpGroup, scan_check
+        >>> from sympy.combinatorics.free_group import free_group
+        >>> F, x, y = free_group("x, y")
+        >>> f = FpGroup(F, [x**2, y**3, (x*y)**4])
+        >>> C = CosetTable(f, [x])
+
+        # scans correctly for (0, x)
+        >>> C.scan_check(0, x)
+        True
+        >>> C.scan(0, x)
+
+        # scans correctly under (0, x**2)
+        >>> C.scan(0, x); C.scan_check(0, x**2)
+        True
+        >>> C.scan(0, x**2); C.scan_check(0, y**3)
+        True
+
+        """
+        # alpha is an integer representing a "coset"
+        # since scanning can be in two cases
+        # 1. for alpha=0 and w in Y (i.e generating set of H)
+        # 2. alpha in omega (set of live cosets), w in R (relators)
+        A_dict = self.A_dict
+        A_dict_inv = self.A_dict_inv
+        f = alpha
+        i = 0
+        r = len(word)
+        b = alpha
+        j = r - 1
+        while i <= j and self.table[f][A_dict[word[i]]] is not None:
+            f = self.table[f][A_dict[word[i]]]
+            i += 1
+        if i > j:
+            return f == b
+        while j >= i and self.table[b][A_dict_inv[word[j]]] is not None:
+            b = self.table[b][A_dict_inv[word[j]]]
+            j -= 1
+        if j < i:
+            # we have an incorrect completed scan with coincidence f ~ b
+            # return False, instead of calling coincidence routine
+            return False
+        else:
```

Comment:
This is the place where we can profitably extend the table by processing a deduction if `i == j`. That will then effectively complete the scan by 'joining' the forwards and backwards scans.